### ISSUE TITLE
Support AsyncValidators on FormGroup

### ISSUE THREAD

**Issue by shaxxx** [joanpablo/reactive_forms#51]
Currently there is no way to provide AsyncValidator on FormGroup wich means it's not possible to do some async validation that takes in a count more fields. 

Take the following example. User needs to enter values in two text fields: "hostname" and "port" and we need to validate if hostname is reachable or not. For that we need both "hostname" and "port" values and we need to raise a validation error if host is not reachable. Putting this in a validator of one control (ie. "hostname") does not triggers when other one changes (ie. "port") and does not update error message.

**Comment by joanpablo** [comment]
Hi @shaxxx,

Thanks for using Reactive Forms and for that issue. I'll provide a solution to solve that problem.

**Comment by shaxxx** [comment]
As far as I could tell it's only a matter of adding extra parameters for AsyncValidators and AsyncDebounceTime to FormGroup.
It kinda worked after adding those parameters, but I was expecting it to call re-validation on non-focused control even when it already has validation errors. That didn't work, but I guess that's another issue.

**Comment by joanpablo** [comment]
@shaxxx could you please write the steps of the behavior that you should expect when adding the async validator that validates Host and Port? 

That will let me write some tests and fully understand your needs.

**Comment by joanpablo** [comment]
Hi @shaxxx,

In Reactive Forms **Async Validators** executes only after all **Sync Validators** are executed and marks the control as **valid**.
That was intentionally implemented that way to avoid requests (to API for example) while the control is not valid with **Sync Validators**. 

Could you please explain me in more details what you are tying to accomplish? I'm sure we are going to find a solution.

I have already add extra parameters to FormGroup (and FormArray also) but before release the version I want to really understand your needs.

Thanks again

**Comment by joanpablo** [comment]
Hi @shaxxx 

I have already release v7.0.8 that includes the Async Validators to **FormGroup** and **FormArray**.

**Comment by shaxxx** [comment]
I'm just writing use case, wait...

**Comment by shaxxx** [comment]
Yes, I figured out that async validators are triggered after regular one, it's logical.
I was able to achieve what I need trough control async validators, very ugly and hackish but works.
Tried with form validators but couldn't make it work as I wanted.
Use case is this. 
- Two controls as stated above, "address" and "port" need to be validated together to check if address:port is available.
- Both control have sync validator required if other one is not empty
- If any of controls is empty don't validate anything
- If both controls have values, and no other validation errors call async validator
- Show validation message on currently focused control
- If unfocused control has that message remove it from the list of errors (keep in mind there may be other errors)

Example:
- form is empty (pristine)
- enter google.com as address (everything is fine, no errors visible but form is invalid due to sync required validator, async validator is not triggered)
- enter port 808 (async validator should now trigger, should display error "Check address and port" on port control)
-modify value in address to "google.co" (async validator should trigger again, CLEAR ERROR from port control and show same error on address control
-modify value in port control to 80 (async validator should trigger again, CLEAR ERROR from address control, form should be valid now, no errors displayed).

Another thing, during async validation FormGroup does not change it's status to validating which makes it hard to disable Submit button during async validation. I made a workaround by putting some custom error at the beginning of async validator function and remove it at the end of function. This makes form invalid during validation and disables my button which is binded to form.valid status.

**Comment by joanpablo** [comment]
Hi @shaxxx,

Thanks for the detailed use case, I will test it now in v7.0.9 to see if it fit the requirement. And I will let you know.

Thanks again.

**Comment by joanpablo** [comment]
Hi @shaxxx,

I've implemented the use case, please tell me if that is what you need.

```dart
import 'package:flutter/material.dart';
import 'package:reactive_forms/reactive_forms.dart';

import 'main.dart';

class TestFormAsync extends StatelessWidget {
  FormGroup _buildForm() {
    return fb.group({
      'address': ['', Validators.required],
      'port': ['', Validators.required],
    }, [], [
      _validServer('address', 'port')
    ]);
  }

  static const String validServer = 'www.google.com:80';
  static const String errorName = 'invalid server';

  AsyncValidatorFunction _validServer(String host, String port) =>
      (AbstractControl control) async {
        FormGroup form = control as FormGroup;
        FormControl<String> hostControl = form.control(host);
        FormControl<String> portControl = form.control(port);
        final error = {
          errorName: {
            host: hostControl.value,
            port: portControl.value,
          }
        };

        // If any of controls is empty don't validate with async validator
        if (hostControl.isNullOrEmpty || portControl.isNullOrEmpty) {
          return Future.value(null);
        }

        // simulates a long server request
        await Future.delayed(Duration(milliseconds: 500), () => null);

        // check if valid URL
        final url = '${hostControl.value}:${portControl.value}';
        if (url.toLowerCase() != validServer.toLowerCase()) {
          // get focused control
          final focusedControl =
              hostControl.hasFocus ? hostControl : portControl;

          // set error on focused control
          focusedControl.setErrors(error);
          focusedControl.markAsTouched();
        } else {
          hostControl.removeError(errorName);
          portControl.removeError(errorName);
        }

        // simulates a request to server
        return Future.value(null);
      };

  @override
  Widget build(BuildContext context) {
    return SampleScreen(
      body: ReactiveFormBuilder(
        form: _buildForm,
        builder: (context, form, child) {
          return Column(
            children: [
              ReactiveTextField(
                formControlName: 'address',
                decoration: InputDecoration(
                  labelText: 'Address',
                ),
              ),
              SizedBox(height: 24.0),
              ReactiveTextField(
                formControlName: 'port',
                decoration: InputDecoration(
                  labelText: 'Port',
                ),
              ),
              SizedBox(height: 24.0),
              ReactiveFormConsumer(
                builder: (context, form, child) {
                  return RaisedButton(
                    child: Text('continue'.toUpperCase()),
                    onPressed: form.valid ? () => {} : null,
                  );
                },
              ),
              SizedBox(height: 24.0),
              Text(
                'Valid Server:',
                style: Theme.of(context).textTheme.caption,
              ),
              Text(
                validServer,
                style: Theme.of(context).textTheme.caption,
              ),
            ],
          );
        },
      ),
    );
  }
}
```

![Screenrecorder-2020-11-05-12-19-25-442](https://user-images.githubusercontent.com/37339180/98275499-b0e53700-1f62-11eb-9610-2a06678bb23a.gif)

**Comment by shaxxx** [comment]
Hi joan,

I'm outside on my mobile but based on the gif it's not scenario I was
describing.
At the end leave port value as invalid and then modify address.
Validation message should move from port to address control.
Maybe your code supports this but I'm unable to try it.
Thank you for your support.

čet, 5. stu 2020. 18:42 Joan Pablo <notifications@github.com> je napisao:

**Comment by joanpablo** [comment]
I see the problem now @shaxxx.

I will study carefully the issue to find a solution.

Thanks.

**Comment by joanpablo** [comment]
Hi @shaxxx,

I have test several ways to accomplish your task. If I let the Async Validator run always (even if sync validations are invalid) then the UI behavior as you expect, but in my opinion that not solve the problem.

If in the UI, I leave **port** value as **invalid** and then modify **address**, then validation message will move from *port* to *address* control. But that will confuse the *user* because if the address is correct but not the port the validation message is visible in address even when it is valid (the port is the one invalid).

So, I would like to propose to you three different better approaches in my opinion.

1-) Add Async Validator to FormGroup and set errors on the FormGroup instead of set errors in each control individually. This will require that you show errors in some other place of the UI, (maybe below both controls).

Here is how it looks:

![Screenrecorder-2020-11-05-14-50-01-990](https://user-images.githubusercontent.com/37339180/98289852-cbc1a680-1f76-11eb-8ee3-161644ad5174.gif)

The complete implementation:

```dart
import 'package:flutter/material.dart';
import 'package:reactive_forms/reactive_forms.dart';

import 'main.dart';

class TestFormAsync extends StatelessWidget {
  FormGroup _buildForm() {
    return fb.group({
      'address': ['', Validators.required],
      'port': ['', Validators.required],
    }, [], [
      _validServer('address', 'port')
    ]);
  }

  static const String validServer = 'www.google.com:80';
  static const String errorName = 'invalid server';

  AsyncValidatorFunction _validServer(String host, String port) =>
      (AbstractControl control) async {
        FormGroup form = control as FormGroup;
        FormControl<String> hostControl = form.control(host);
        FormControl<String> portControl = form.control(port);
        final error = {
          errorName: {
            host: hostControl.value,
            port: portControl.value,
          }
        };

        // If any of controls is empty don't validate with async validator
        if (hostControl.isNullOrEmpty || portControl.isNullOrEmpty) {
          return Future.value(null);
        }

        final url = '${hostControl.value.trim()}:${portControl.value.trim()}';

        // simulates a long server request
        return Future.delayed(
          Duration(milliseconds: 500),
          () => url.toLowerCase() != validServer.toLowerCase() ? error : null,
        );
      };

  InputDecoration errorDecoration(context) => InputDecoration(
        labelStyle: TextStyle(color: Colors.red),
        focusedBorder: Theme.of(context).inputDecorationTheme.border.copyWith(
              borderSide: BorderSide(
                color: Colors.red,
                width: 2.0,
              ),
            ),
        enabledBorder: Theme.of(context).inputDecorationTheme.border.copyWith(
              borderSide: BorderSide(
                color: Colors.red,
              ),
            ),
      );

  @override
  Widget build(BuildContext context) {
    return SampleScreen(
      body: ReactiveFormBuilder(
        form: _buildForm,
        builder: (context, form, child) {
          return Column(
            children: [
              ReactiveFormConsumer(builder: (_, form, child) {
                return Row(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Expanded(
                      flex: 2,
                      child: ReactiveTextField(
                        formControlName: 'address',
                        decoration: form.hasError(errorName)
                            ? errorDecoration(context)
                                .copyWith(labelText: 'Address')
                            : InputDecoration(labelText: 'Address'),
                      ),
                    ),
                    SizedBox(width: 10.0),
                    Expanded(
                      child: ReactiveTextField(
                        formControlName: 'port',
                        decoration: form.hasError(errorName)
                            ? errorDecoration(context)
                                .copyWith(labelText: 'Port')
                            : InputDecoration(labelText: 'Port'),
                      ),
                    ),
                  ],
                );
              }),
              ReactiveFormConsumer(
                builder: (context, form, child) {
                  if (form.pending) {
                    return Padding(
                      padding: const EdgeInsets.all(8.0),
                      child: CircularProgressIndicator(),
                    );
                  }

                  return Visibility(
                    visible: form.hasError(errorName),
                    child: Padding(
                      padding: const EdgeInsets.all(8.0),
                      child: Text(
                        'The address and port are not valid!',
                        style: Theme.of(context)
                            .textTheme
                            .bodyText2
                            .copyWith(color: Colors.red),
                      ),
                    ),
                  );
                },
              ),
              SizedBox(height: 24.0),
              ReactiveFormConsumer(
                builder: (context, form, child) {
                  return RaisedButton(
                    child: Text('continue'.toUpperCase()),
                    onPressed: form.valid ? () => {} : null,
                  );
                },
              ),
              SizedBox(height: 24.0),
              Text(
                'Valid Server:',
                style: Theme.of(context).textTheme.caption,
              ),
              Text(
                validServer,
                style: Theme.of(context).textTheme.caption,
              ),
            ],
          );
        },
      ),
    );
  }
}
```

In case the Form has several others fields you can create a subgroup just for Address and Port.

2-) The second approach is a more simplistic implementation. You could use only one **ReactiveTextField** and let the user type the Address and Port in the same input by supplying a **mask** [TextInputFormatter](https://api.flutter.dev/flutter/services/TextInputFormatter-class.html). This will require you to implement that formatter or use some plugin.

There are several plugin based in TextInputFormatter, here are some of them:
- [mask_text_input_formatter](https://pub.dev/packages/mask_text_input_formatter).
- [text_input_mask_formatter](https://pub.dev/packages/text_input_mask_formatter).
- [currency_text_input_formatter](https://pub.dev/packages/currency_text_input_formatter).

This second approach lead to another question, do you want to keep Address and Port in separated models (separated FormControl) or in the same? If you want to separate data you have a few options like Bind the input with a FormControl of type array.

3-) The third approach is more sophisticated and require you to implement a Custom Reactive Widget.

**Comment by shaxxx** [comment]
In the end I decided to go with your first solution, maybe not what I initially intended but it kinda makes sense to keep the error on the control that originally caused it. Plus your validator is way prettier than code I wrote and can be re-used (I actually have to check address:port for local and remote server).
New version works great with debounce values, submit button reacts instantly.
Great work, and once again thank you for your support.
Keep up the good work.

**Comment by joanpablo** [comment]
Thanks for your kind words @shaxxx, I'm glad to help you. 

On last advice, In the above example I use to many **ReactiveFormConsumer**, one for button, other for the error message and other for the inputs, this could be more simple if using just one ReactiveFormConsumer because the example was very simple.

Feel free to open other issues if you have other problems.

Thanks.
